File: solutions/94. Binary Tree Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        
        res=[]
       
        curr=root
        q = deque([])
        
        while  curr or len(q)>0:
           
            while curr:
                q.append(curr)
                curr=curr.left
            curr=q.pop()
            res.append(curr.val)
            curr=curr.right
            
        return res
        
        
        '''
        self.result=[]
        
        def helper(root):
            if not root:
                return 
            helper(root.left)
            self.result.append(root.val)
            helper(root.right)
         
        helper(root)
        return self.result
        '''

File: solutions/94. Binary Tree Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        
        List<Integer> res= new ArrayList<>();
        Stack<TreeNode> stack= new Stack<>();
        TreeNode node=root;
        
        while (node!=null || !stack.isEmpty()) {
            while (node!=null) {
                //System.out.println(node.val);
                stack.push(node);
                node=node.left;
            }
            
            node =stack.pop();
            //System.out.println(node);
            res.add(node.val);
            node=node.right;
        }
        
        return res;
    }
        
        
        
        
        
        
        
        
        
        
       /*
       List<Integer> res= new ArrayList<>();
       helper(root,res);
       return res;
    }
